app/assets/javascripts/angular/controllers/frame_ctrl.js

```diff
@@ -36,33 +36,37 @@ angular.module("Prometheus.controllers").controller('FrameCtrl', ["$scope",
   };
 
   function buildFrameURL(url) {
-    var parser = document.createElement('a');
+    var parser = document.createElement("a");
     parser.href = url;
-    var queryStringComponents = parser.search.substring(1).split('&');
+    var queryStringComponents = parser.search.substring(1).split("&");
     if ($scope.frame.graphite) {
-      if (url.indexOf("until") === -1) {
-        queryStringComponents.push("until=now");
-      }
+      ["height=0", "width=0", "from=-1h", "until=now"].forEach(function(e) {
+        // Unshift elements to the front of the query string. If `from`
+        // already exists, it will not be clobbered by these values.
+        queryStringComponents.unshift(e);
+      });
```

Comment:
This code becomes more and more complicated, with unclear side effects which require comments to explain what's going on. I might be missing something, but maybe rewriting the whole function including `setDimension` could help here? A first idea:

```javascript
function buildFrameURL(url) {
  var parser = document.createElement('a');
  parser.href = url;
  var queryStringComponents = parser.search.substring(1).split('&');
  if ($scope.frame.graphite) {
    var fields = {};
    queryStringComponents.forEach(function(f) {
      var s = f.split("=");
      fields[s.shift()] = s.join('=');
    });
    
    fields['height'] = $scope.frameHeight().height;
    fields['width'] = $scope.frameHeight().height / $scope.aspectRatio;
    fields['from'] = GraphiteTimeConverter.graphiteFrom($scope.frame.range, $scope.frame.endTime);
    fields['until'] = GraphiteTimeConverter.graphiteUntil($scope.frame.endTime);
    fields['bgcolor'] = '%23191919'

    // join fields
  }

  // ...
}
```

That approach would ensure that the 4 parameters always get set correctly, regardless of what parameters are set already. I also believe that not using this switch statement makes the code more comprehensible.

Ideally, the code to parse/construct an URL should be separated from the business logic to construct a graphite URL.